### PR TITLE
Multisig: State should ONLY have ID address -> create accounts if they do NOT exist

### PR DESCRIPTION
**Pull request by aarshkshah1992**
For #730.

@anorth We should move the `resolveToIdAddr` func to a common location so the `Verifreg` and `PaymentChannel` actors can use it too.

### CODE REVIEW

**Comment by aarshkshah1992** [comment]
@wadeAlexC.

**Comment by wadeAlexC** [comment]
These changes make me think the runtime should expose a "ForceResolve" method that automatically creates an account if it doesn't exist.

Either that, or the concept of non-existent accounts could just be abstracted away entirely, and accounts are always created if they don't already exist. i.e. things like `rt.DeleteActor(beneficiary)` would create `beneficiary` if it doesn't exist.

**Comment by wadeAlexC** [comment]
What would be really cool is if this were abstracted in a way s.t. ALL addresses were always ID-addresses within code, and a reverse ID -> BLS/SECP/ACTOR lookup existed.

**Comment by wadeAlexC** [comment]
I would like to see these changes decouple ID address resolution from Multisig-specific functionality.

`isSigner` should assert that it receives an ID address; resolution should be an individual, explicit action taken as soon as an address is referenced for the first time.

File: actors/builtin/multisig/multisig_actor.go [hunk 1]
```diff
@@ -302,17 +320,20 @@ func (a Actor) RemoveSigner(rt vmr.Runtime, params *RemoveSignerParams) *adt.Emp
...
 			rt.Abortf(exitcode.ErrForbidden, "cannot remove only signer")
 		}
 
 		newSigners := make([]addr.Address, 0, len(st.Signers))
+		// signers have already been resolved
 		for _, s := range st.Signers {
-			if !isAddressEqual(rt.ResolveAddress, s, params.Signer) {
+			if resolved != s {
 				newSigners = append(newSigners, s)
 			}
```
Comment (wadeAlexC) [inline actors/builtin/multisig/multisig_actor.go:338]: Do we still need to iterate over st.Signers? The !isSigner check should be functionally identical.

File: actors/builtin/multisig/multisig_actor.go [hunk 2]
```diff
@@ -344,21 +365,25 @@ func (a Actor) SwapSigner(rt vmr.Runtime, params *SwapSignerParams) *adt.EmptyVa
...
 		newSigners := make([]addr.Address, 0, len(st.Signers))
 		for _, s := range st.Signers {
-			if !isAddressEqual(rt.ResolveAddress, s, params.From) {
+			if s != fromResolved {
 				newSigners = append(newSigners, s)
 			}
 		}
-		newSigners = append(newSigners, params.To)
+		newSigners = append(newSigners, toResolved)
 		st.Signers = newSigners
```
Comment (wadeAlexC) [inline actors/builtin/multisig/multisig_actor.go:387]: Same comment here. I dont think we need to iterate over `st.Signers` anymore.

File: actors/builtin/multisig/multisig_actor.go [hunk 3]
```diff
@@ -384,7 +409,7 @@ func (a Actor) ChangeNumApprovalsThreshold(rt vmr.Runtime, params *ChangeNumAppr
 	return nil
 }
 
-func (a Actor) approveTransaction(rt vmr.Runtime, txnID TxnID, txn *Transaction) (bool, []byte, exitcode.ExitCode) {
+func (a Actor) approveTransaction(rt vmr.Runtime, resolvedApprover addr.Address, txnID TxnID, txn *Transaction) (bool, []byte, exitcode.ExitCode) {
```
Comment (wadeAlexC) [inline actors/builtin/multisig/multisig_actor.go:412]: Should AssertMsg that `resolvedApprover` is an ID address

File: actors/builtin/multisig/multisig_actor.go [hunk 4]
```diff
@@ -399,9 +424,9 @@ func (a Actor) approveTransaction(rt vmr.Runtime, txnID TxnID, txn *Transaction)
 		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load pending transactions")
 
 		// update approved on the transaction
-		txn.Approved = append(txn.Approved, rt.Message().Caller())
-		 err = ptx.Put(txnID, txn)
-		 builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to put transaction %v for approval", txnID)
+		txn.Approved = append(txn.Approved, resolvedApprover)
+		err = ptx.Put(txnID, txn)
+		builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to put transaction %v for approval", txnID)
```
Comment (wadeAlexC) [inline actors/builtin/multisig/multisig_actor.go:429]: Should the "abort duplicate approval" check above use `resolveApprover` as well?

File: actors/builtin/multisig/multisig_actor.go [hunk 5]
```diff
@@ -476,33 +501,50 @@ func executeTransactionIfApproved(rt vmr.Runtime, st State, txnID TxnID, txn *Tr
...
 
-	for _, ap := range st.Signers {
-		signerResolved := resolve(resolveFunc, ap)
-		if signerResolved == candidateResolved {
-			return true
+	// signer addresses have already been resolved
+	for _, ap := range signers {
+		if ap == candidateResolved {
+			return true, candidateResolved, nil
 		}
```
Comment (wadeAlexC) [inline actors/builtin/multisig/multisig_actor.go:517]: `ap`? Not sure what this stands for. Maybe use `signer` instead?

File: actors/builtin/multisig/multisig_actor.go [hunk 6]
```diff
@@ -476,33 +501,50 @@ func executeTransactionIfApproved(rt vmr.Runtime, st State, txnID TxnID, txn *Tr
...
-	if resolved.Protocol() != addr.ID {
-		idAddr, found := resolveFunc(resolved)
-		if found {
-			resolved = idAddr
-		}
+func resolveToIDAddr(address addr.Address, resolveFunc AddressResolveFunc, sendFunc SendFunc) (addr.Address, error) {
+	// if it's already an ID address, we can return immediately
+	if address.Protocol() == addr.ID {
+		return address, nil
 	}
```
Comment (wadeAlexC) [inline actors/builtin/multisig/multisig_actor.go:527]: This short-circuit doesn't guarantee that `address` exists if it's already an ID-address.

Comment (wadeAlexC) [reply actors/builtin/multisig/multisig_actor.go:527]: Plus, `resolveFunc` performs the same short-circuit internally, so I think this branch is just redundant.

File: actors/builtin/multisig/multisig_actor.go (hunk 3 above)
Comment (aarshkshah1992) [reply actors/builtin/multisig/multisig_actor.go:412]: This is done.

File: actors/builtin/multisig/multisig_actor.go (hunk 6 above)
Comment (aarshkshah1992) [reply actors/builtin/multisig/multisig_actor.go:527]: I agree. This has been removed.

File: actors/builtin/multisig/multisig_actor.go (hunk 5 above)
Comment (aarshkshah1992) [reply actors/builtin/multisig/multisig_actor.go:517]: This is done.

File: actors/builtin/multisig/multisig_actor.go (hunk 4 above)
Comment (aarshkshah1992) [reply actors/builtin/multisig/multisig_actor.go:429]: This is done.

File: actors/builtin/multisig/multisig_actor.go (hunk 1 above)
Comment (aarshkshah1992) [reply actors/builtin/multisig/multisig_actor.go:338]: I am not sure what you mean here, This iteration is to get the Signers slice without the signer we need to remove.
Can you elaborate on what you mean here ?

File: actors/builtin/multisig/multisig_actor.go (hunk 2 above)
Comment (aarshkshah1992) [reply actors/builtin/multisig/multisig_actor.go:387]: Same comment as above.

File: actors/builtin/multisig/multisig_actor.go (hunk 4 above)
Comment (aarshkshah1992) [reply actors/builtin/multisig/multisig_actor.go:438]: Also added a test to catch this.

**Comment by aarshkshah1992** [comment]
@wadeAlexC 

> I would like to see these changes decouple ID address resolution from Multisig-specific functionality.
> 
> isSigner should assert that it receives an ID address; resolution should be an individual, explicit action taken as soon as an address is referenced for the first time.

This is done.

> 
> These changes make me think the runtime should expose a "ForceResolve" method that automatically creates an account if it doesn't exist.
> 
> Either that, or the concept of non-existent accounts could just be abstracted away entirely, and accounts are always created if they don't already exist. i.e. things like rt.DeleteActor(beneficiary) would create beneficiary if it doesn't exist.

While I think it's a bit too late to make any change to the `Runtime` Impl, I'll leave it to @anorth to comment on this.

File: actors/builtin/multisig/multisig_actor.go [hunk 7]
```diff
@@ -141,6 +146,9 @@ func (a Actor) Propose(rt vmr.Runtime, params *ProposeParams) *ProposeReturn {
 	rt.ValidateImmediateCallerType(builtin.CallerTypesSignable...)
 	callerAddr := rt.Message().Caller()
 
+	resolvedCaller, err := resolveToIDAddr(callerAddr, rt.ResolveAddress, rt.Send)
+	builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to resolved address %v", callerAddr)
```
Comment (anorth) [inline actors/builtin/multisig/multisig_actor.go:150]: Unnecessary, the caller is always an ID address.

File: actors/builtin/multisig/multisig_actor.go (hunk 3 above)
Comment (anorth) [reply actors/builtin/multisig/multisig_actor.go:419]: It was unnecessary to resolve the caller in the first place, so also unnecessary to pass it as a parameter here or assert anything.

File: actors/builtin/multisig/multisig_actor.go [hunk 8]
```diff
@@ -476,33 +510,40 @@ func executeTransactionIfApproved(rt vmr.Runtime, st State, txnID TxnID, txn *Tr
...
 }
 
-func resolve(resolveFunc AddressResolveFunc, address addr.Address) addr.Address {
-	resolved := address
-	if resolved.Protocol() != addr.ID {
-		idAddr, found := resolveFunc(resolved)
-		if found {
-			resolved = idAddr
-		}
+func resolveToIDAddr(address addr.Address, resolveFunc AddressResolveFunc, sendFunc SendFunc) (addr.Address, error) {
```
Comment (anorth) [inline actors/builtin/multisig/multisig_actor.go:527]: Move this to `builtin/shared.go`. Also it's ok for it to just take Runtime as a parameter, since it' always called at the top level and obviously needs it.

File: actors/builtin/multisig/multisig_test.go [hunk 9]
```diff
@@ -31,15 +31,15 @@ func TestConstruction(t *testing.T) {
...
 
 	bob := tutil.NewIDAddr(t, 102)
 	bobNonId := tutil.NewBLSAddr(t, 2)
 
 	charlie := tutil.NewIDAddr(t, 103)
 
-	builder := mock.NewBuilder(context.Background(), receiver).WithCaller(builtin.InitActorAddr, builtin.InitActorCodeID)
-
 	t.Run("simple construction", func(t *testing.T) {
+		builder := mock.NewBuilder(context.Background(), receiver).WithCaller(builtin.InitActorAddr, builtin.InitActorCodeID)
```
Comment (anorth) [inline actors/builtin/multisig/multisig_test.go:42]: Why did you move and duplicate this line in all the cases? They all need it, and if something needs something special, you can shadow it or use another name.

File: actors/builtin/multisig/multisig_actor.go (hunk 2 above)
Comment (wadeAlexC) [reply actors/builtin/multisig/multisig_actor.go:394]: I guess my question is - can you just do `st.Signers = append(st.Signers, toResolved)`?

Comment (wadeAlexC) [reply actors/builtin/multisig/multisig_actor.go:394]: Iterating over happens twice in this method - once in isSigner, and once here. We only need to iterate a single time, I think!

Comment (wadeAlexC) [reply actors/builtin/multisig/multisig_actor.go:394]: Nevermind, this doesn't work for SwapSigner. But I think it does work for AddSigner!

File: actors/builtin/multisig/multisig_actor.go (hunk 7 above)
Comment (aarshkshah1992) [reply actors/builtin/multisig/multisig_actor.go:150]: This is fixed everywhere.

File: actors/builtin/multisig/multisig_actor.go (hunk 8 above)
Comment (aarshkshah1992) [reply actors/builtin/multisig/multisig_actor.go:527]: This is done.

File: actors/builtin/multisig/multisig_test.go (hunk 9 above)
Comment (aarshkshah1992) [reply actors/builtin/multisig/multisig_test.go]: Apologies, this is fixed.

**Comment by aarshkshah1992** [comment]
@anorth Have made all the changes. Please take a look.

**Comment by wadeAlexC** [comment]
LGTM

File: actors/builtin/multisig/multisig_actor.go [hunk 10]
```diff
@@ -105,7 +110,7 @@ func (a Actor) Constructor(rt vmr.Runtime, params *ConstructorParams) *adt.Empty
 	}
 
 	var st State
-	st.Signers = params.Signers
+	st.Signers = resolvedAddrs
```
Comment (anorth) [inline actors/builtin/multisig/multisig_actor.go:113]: Nit: resolvedSigners?

File: actors/builtin/multisig/multisig_actor.go [hunk 11]
```diff
@@ -149,7 +154,8 @@ func (a Actor) Propose(rt vmr.Runtime, params *ProposeParams) *ProposeReturn {
 	var st State
 	var txn *Transaction
 	rt.State().Transaction(&st, func() {
-		if !isSigner(rt.ResolveAddress, &st, callerAddr) {
+		proposerIsSigner := isSigner(callerAddr, st.Signers)
+		if !proposerIsSigner {
```
Comment (anorth) [inline actors/builtin/multisig/multisig_actor.go:158]: Nit: I would fold these back to one line; the predicate code is already pretty clear. Possibly rename `callerAddr` for even more clarity.

File: actors/builtin/multisig/multisig_actor.go [hunk 12]
```diff
@@ -299,20 +312,24 @@ type RemoveSignerParams struct {
 func (a Actor) RemoveSigner(rt vmr.Runtime, params *RemoveSignerParams) *adt.EmptyValue {
 	// Can only be called by the multisig wallet itself.
 	rt.ValidateImmediateCallerIs(rt.Message().Receiver())
+	resolvedSigner, err := builtin.ResolveToIDAddr(rt, params.Signer)
```
Comment (anorth) [inline actors/builtin/multisig/multisig_actor.go:315]: resolvedOldSigner?

File: actors/builtin/multisig/multisig_actor.go [hunk 13]
```diff
@@ -385,10 +410,13 @@ func (a Actor) ChangeNumApprovalsThreshold(rt vmr.Runtime, params *ChangeNumAppr
 }
 
 func (a Actor) approveTransaction(rt vmr.Runtime, txnID TxnID, txn *Transaction) (bool, []byte, exitcode.ExitCode) {
+
```
Comment (anorth) [inline actors/builtin/multisig/multisig_actor.go:413]: Nit: unnecessary blank line

File: actors/builtin/multisig/multisig_actor.go [hunk 14]
```diff
@@ -476,35 +504,20 @@ func executeTransactionIfApproved(rt vmr.Runtime, st State, txnID TxnID, txn *Tr
 }
 
 type AddressResolveFunc func(address addr.Address) (resolved addr.Address, found bool)
+type SendFunc func(toAddr addr.Address, methodNum abi.MethodNum, params vmr.CBORMarshaler, value abi.TokenAmount) (vmr.SendReturn, exitcode.ExitCode)
```
Comment (anorth) [inline actors/builtin/multisig/multisig_actor.go:507]: I think this is redundant now

File: actors/builtin/shared.go [hunk 15]
```diff
@@ -54,3 +56,27 @@ type MinerAddrs struct {
...
+	// if we are able to resolve it to an ID address, return the resolved address
+	idAddr, found := rt.ResolveAddress(address)
+	if found {
+		return idAddr, nil
+	}
+
+	// send 0 balance to the account so an ID address for it is created and then try to resolve
+	_, code := rt.Send(address, MethodSend, nil, abi.NewTokenAmount(0))
+	if !code.IsSuccess() {
+		return address, fmt.Errorf("failed to send zero balance to account %v, got code %v", address, code)
```
Comment (anorth) [inline actors/builtin/shared.go:72]: `code` here is an `ExitCode` which implements error, so we should return a thing that will result in propagating that exit code. I think you want something like `code.Wrapf(msg, args)`.
